### PR TITLE
#163426544 implement role administration

### PR DESCRIPTION
**Pull request by bobbylemm**
#### What does this PR do?
Implement role and permissions functionality.
#### Description of task to be completed?
Have the following endpoints working.
`post /api/v1/admin/role`
#### How should this be manually tested?
After cloning the repo, run `npm install` and then `npm test`.

- Using Postman test the endpoints stated above.
key: provide a `username` and `role` for this endpoint.
key: the roles accepted are `user` and `admin`

#### Any background context you want to provide?
In your `.env` file, state a `SUPER_ADMIN_EMAIL` and `SUPER_ADMIN_PASSWORD` 
#### Relevant PT story ID?
[#163426544](https://www.pivotaltracker.com/story/show/163426544)

### CODE REVIEW

**Review by ozone4real — DISMISSED** [review]
LGTM

**Review by biolabalo — DISMISSED** [review]
LGTM

**Review by omenkish — DISMISSED** [review]

**Review by EfeAgare — DISMISSED** [review]

**Review by omenkish — DISMISSED** [review]

**Review by ozone4real — DISMISSED** [review]
LGTM

**Review by biolabalo — DISMISSED** [review]

File: server/controllers/userController.js [hunk 1]
```diff
@@ -171,28 +173,28 @@ class UserController extends BaseHelper {
...
-        if (!user) {
-          return errorMessage(res, 404, 'User not found');
-        }
-        return res.status(200).json({
-          profile: user.dataValues
-        });
+      });
+      if (!user) {
+        return errorMessage(res, 404, 'User not found');
+      }
```
Comment (p-funky) [inline server/controllers/userController.js]: Why not use the `checkIfDataExist` method from the baseHelper class?

File: server/controllers/RoleAssignController.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+   * @returns {Object} it returns either the updated role of the user or an error
+   *
+   * @memberOf RoleAssignController
+   */
+  static async assignRole(req, res, next) {
+    const { username, role } = req.body;
+    try {
+      const existingRole = await usersRepo.role(username, res);
+      if (existingRole === role) {
+        return errorMessage(res, 401, `this user is already an ${role}`);
```
Comment (p-funky) [inline server/controllers/RoleAssignController.js]: For example, `this user is already an user`? 😳

**Review by EfeAgare — DISMISSED** [review]

**Review by ozone4real — DISMISSED** [review]
LGTM

File: server/controllers/RoleAssignController.js (hunk 2 above)
Comment (p-funky) [reply server/controllers/RoleAssignController.js]: Or is it that you can only update the roles to `admin`?

File: server/controllers/userController.js (hunk 1 above)
Comment (bobbylemm) [reply server/controllers/userController.js]: this seemed more readable to me, should I change it

File: server/controllers/RoleAssignController.js (hunk 2 above)
Comment (bobbylemm) [reply server/controllers/RoleAssignController.js]: @p-funky should I have a/an instead

File: server/controllers/userController.js (hunk 1 above)
Comment (p-funky) [reply server/controllers/userController.js]: It's fine though but in the spirit of consistency with what's already in use within this controller, I'll request you change it please.

**Comment by p-funky** [comment]
1. What differentiates an `admin` from a `regular user`?
2. How many roles actually exist?
3. Can any role name be saved in this roles column?

cc @bobbylemm @omenkish @biolabalo @primuse @EfeAgare

**Comment by bobbylemm** [comment]
@p-funky what differentiates an `admin` from a regular `user` is majorly the `role` column on the user's table. 
1)An admin can delete a user, see the report written on an article, take action on that report e.g to delete the article.
2) there is a `super admin` `admin` and `user`
3) Not just any role name can be saved in the roles column

**Comment by p-funky** [comment]
1) An admin `can or should be able to` ... I don't see any of those implemented yet.
2) There should only `super admin` `admin` and `user` roles but seeing it is simply a string in the user model, this means I can send in any role I please and it will accept it.

File: server/controllers/RoleAssignController.js (hunk 2 above)
Comment (p-funky) [reply server/controllers/RoleAssignController.js]: Why not "The role of this user is already ${role}"

File: env.sample [hunk 3]
```diff
@@ -0,0 +1,7 @@
+DATABASE_URL_DEV=development_DB_URL_string
+GOOGLE_CLIENT_ID=areallylongstringofnumbersandalphabets
+GOOGLE_CLIENT_SECRET=anotherlongstringofnumbersandalphabets
+FACEBOOK_CLIENT_ID=anotherlongstringofnumbersandalphabets
+FACEBOOK_CLIENT_SECRET=anotherlongstringofnumbersandalphabets
+TWITTER_CLIENT_ID=anotherlongstringofnumbersandalphabets
+TWITTER_CLIENT_SECRET=anotherlongstringofnumbersandalphabets
```
Comment (p-funky) [inline env.sample]: This file needs to be updated with the new env variables

File: server/controllers/userController.js [hunk 4]
```diff
@@ -134,6 +133,9 @@ class UserController extends BaseHelper {
         where: { email },
         attributes: ['name', 'username', 'email', 'password', 'role', 'isVerified', 'id']
       });
+      if (!newUser) {
+        return errorMessage(res, 404, 'error logging in');
+      }
```
Comment (p-funky) [inline server/controllers/userController.js]: Let's make use of `checkIfDataExist` here as well.

File: server/db/repositories/base.js [hunk 5]
```diff
@@ -53,21 +53,19 @@ class BaseRepository {
...
+          where: { username },
+          returning: true
+        }
+      );
+      if (updatedRole) {
+        return res.status(200).json({
+          message: 'successfully updated user role',
+          updatedUser: updatedRole[1][0].dataValues
+        });
+      }
```
Comment (p-funky) [inline server/db/repositories/base.js:68]: Lol. Did not want to bother you further with this before you accuse me of stressing you unnecessarily. 🙂🙃

**Review by p-funky — APPROVED** [review]

**Review by biolabalo — APPROVED** [review]

**Review by EfeAgare — APPROVED** [review]

**Pull request merged**